external/esp_idf_port/esp_lwip/dhcpserver.c

```diff
@@ -0,0 +1,1362 @@
+	}
+
+	ip_addr_t ip_temp = IPADDR4_INIT(0x0);
+	ip4_addr_set(ip_2_ip4(&ip_temp), &broadcast_dhcps);
+	SendOffer_err_t = udp_sendto(pcb_dhcps, p, &ip_temp, DHCPS_CLIENT_PORT);
+	DHCPS_LOG("dhcps: send_offer>>udp_sendto result %x\n", SendOffer_err_t);
+
+	if (p->ref != 0) {
+#if DHCPS_DEBUG
+		DHCPS_LOG("udhcp: send_offer>>free pbuf\n");
+#endif
+		pbuf_free(p);
+	}
+}
+
+/******************************************************************************
+ * FunctionName : send_nak
+ * Description  : DHCP message NACK Response
+ * Parameters   : m -- DHCP message info
+ * Returns      : none
+*******************************************************************************/
+static void send_nak(struct dhcps_msg *m, u16_t len)
+{
+	u8_t *end;
+	struct pbuf *p, *q;
+	u8_t *data;
+	u16_t cnt = 0;
+	u16_t i;
+	err_t SendNak_err_t;
+	create_msg(m);
+
+	end = add_msg_type(&m->options[4], DHCPNAK);
+	end = add_end(end);
+
+	p = dhcps_pbuf_alloc(len);
+#if DHCPS_DEBUG
+	DHCPS_LOG("udhcp: send_nak>>p->ref = %d\n", p->ref);
+#endif
+
+	if (p != NULL) {
+
+#if DHCPS_DEBUG
+		DHCPS_LOG("dhcps: send_nak>>pbuf_alloc succeed\n");
+		DHCPS_LOG("dhcps: send_nak>>p->tot_len = %d\n", p->tot_len);
+		DHCPS_LOG("dhcps: send_nak>>p->len = %d\n", p->len);
+#endif
+		q = p;
+
+		while (q != NULL) {
+			data = (u8_t *) q->payload;
+
+			for (i = 0; i < q->len; i++) {
+				data[i] = ((u8_t *) m)[cnt++];
+#if DHCPS_DEBUG
+				DHCPS_LOG("%02x ", data[i]);
+
+				if ((i + 1) % 16 == 0) {
+					DHCPS_LOG("\n");
+				}
+#endif
+			}
+
+			q = q->next;
+		}
+	} else {
+
+#if DHCPS_DEBUG
+		DHCPS_LOG("dhcps: send_nak>>pbuf_alloc failed\n");
+#endif
+		return;
+	}
+
+	ip_addr_t ip_temp = IPADDR4_INIT(0x0);
+	ip4_addr_set(ip_2_ip4(&ip_temp), &broadcast_dhcps);
+	SendNak_err_t = udp_sendto(pcb_dhcps, p, &ip_temp, DHCPS_CLIENT_PORT);
+	DHCPS_LOG("dhcps: send_nak>>udp_sendto result %x\n", SendNak_err_t);
+
+	if (p->ref != 0) {
+#if DHCPS_DEBUG
+		DHCPS_LOG("udhcp: send_nak>>free pbuf\n");
+#endif
+		pbuf_free(p);
+	}
+}
+
+/******************************************************************************
+ * FunctionName : send_ack
+ * Description  : DHCP message ACK Response
+ * Parameters   : m -- DHCP message info
+ * Returns      : none
+*******************************************************************************/
+static void send_ack(struct dhcps_msg *m, u16_t len)
+{
+	u8_t *end;
+	struct pbuf *p, *q;
+	u8_t *data;
+	u16_t cnt = 0;
+	u16_t i;
+	err_t SendAck_err_t;
+	create_msg(m);
+
+	end = add_msg_type(&m->options[4], DHCPACK);
+	end = add_offer_options(end);
+	end = add_end(end);
+
+	p = dhcps_pbuf_alloc(len);
+#if DHCPS_DEBUG
+	DHCPS_LOG("udhcp: send_ack>>p->ref = %d\n", p->ref);
+#endif
+
+	if (p != NULL) {
+
+#if DHCPS_DEBUG
+		DHCPS_LOG("dhcps: send_ack>>pbuf_alloc succeed\n");
+		DHCPS_LOG("dhcps: send_ack>>p->tot_len = %d\n", p->tot_len);
+		DHCPS_LOG("dhcps: send_ack>>p->len = %d\n", p->len);
+#endif
+		q = p;
+
+		while (q != NULL) {
+			data = (u8_t *) q->payload;
```

Comment:
please check all the lines. remove the speces.

Reply:
done